crates/edr_solidity/src/vm_tracer.rs

```diff
@@ -0,0 +1,276 @@
+        }
+    }
+}
+
+impl VmTracer {
+    /// Creates a new [`VmTracer`].
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Returns a reference to the last top-level message trace.
+    /// # Panics
+    /// This function panics if executed concurrently with [`Self::observe`].
+    pub fn get_last_top_level_message_trace(&self) -> Option<MessageTrace> {
+        self.message_traces
+            .last()
+            .map(|x| RefCell::borrow(x).clone())
+    }
+
+    /// Returns a reference to the last top-level message trace.
+    /// The reference is only being mutated for the duration of
+    /// [`Self::observe`] call.
+    pub fn get_last_top_level_message_trace_ref(&self) -> Option<&MessageTraceRefCell> {
+        self.message_traces.first()
+    }
+
+    /// Retrieves the last error that occurred during the tracing process.
+    pub fn get_last_error(&self) -> Option<&'static str> {
+        self.last_error
+    }
+
+    /// Observes a trace, collecting information about the execution of the EVM.
+    pub fn observe(&mut self, trace: &edr_evm::trace::Trace) {
+        for msg in &trace.messages {
+            match msg.clone() {
+                edr_evm::trace::TraceMessage::Before(before) => {
+                    self.add_before_message(before);
+                }
+                edr_evm::trace::TraceMessage::Step(step) => {
+                    self.add_step(step);
+                }
+                edr_evm::trace::TraceMessage::After(after) => {
+                    self.add_after_message(after.execution_result);
+                }
+            }
+        }
+    }
+
+    fn should_keep_tracing(&self) -> bool {
+        self.last_error.is_none()
+    }
+
+    fn add_before_message(&mut self, message: BeforeMessage) {
+        if !self.should_keep_tracing() {
+            return;
+        }
+
+        let trace: MessageTrace;
+
+        if message.depth == 0 {
+            self.message_traces.clear();
+            self.tracing_steps.clear();
+        }
+
+        if message.to.is_none() {
+            let create_trace = CreateMessageTrace {
+                base: BaseEvmMessageTrace {
+                    base: BaseMessageTrace {
+                        depth: message.depth,
+                        value: message.value,
+                        exit: ExitCode::Success,
+                        return_data: Bytes::new(),
+                        gas_used: 0,
+                    },
+                    code: message.data,
+                    steps: vec![],
+                    number_of_subtraces: 0,
+                    // this was not in the original code - assumed to be None/undefined?
+                    bytecode: None,
+                },
+
+                deployed_contract: None,
+            };
+
+            trace = MessageTrace::Create(create_trace);
+        } else {
+            // TODO: Make this nicer and make sure the U160 logic/precompile comparison
+            // logic works
```

Comment:
Todo. Are you planning to resolve this before merging?

If not, could you please create a GitHub issue for this and link to its URL here?